chainermn/functions/batch_normalization.py

```diff
@@ -142,192 +121,30 @@ def get_communication_backend(comm, communication_backend='auto'):
     return selected_communication_backend
 
 
-class MultiNodeBatchNormalizationFunction(function.Function):
-
+class MultiNodeBatchNormalizationFunction(BatchNormalization):
     def __init__(self, comm, eps=2e-5, mean=None, var=None, decay=0.9,
                  communication_backend='auto'):
-        chainer.utils.experimental(
-            'chainermn.functions.MultiNodeBatchNormalizationFunction')
-
+        super().__init__(eps, mean, var, decay)
         self.comm = comm
-        self.running_mean = mean
-        self.running_var = var
-
-        # Note: cuDNN v5 requires that eps be greater than 1e-5. Otherwise, an
-        # error will occur.
-        # See CUDNN_BN_MIN_EPSILON value in cudnn.h to verify minimum allowable
-        # value.
-        self.eps = eps
-        if chainer.should_use_cudnn('>=auto'):
-            if eps < 1e-5:
-                msg = 'cuDNN does not allow an eps value less than 1e-5.'
-                raise RuntimeError(msg)
-        self.mean_cache = None
-        self.decay = decay
-
-        selected_communication_backend = \
-            get_communication_backend(comm, communication_backend)
+        self.communication_backend = communication_backend
 
-        if selected_communication_backend == 'nccl':
-            self._backend = _NcclBackend(comm)
+    def _bn_backend_selector(self, xp):
```

Comment:
How about passing a "selector" object to `BatchNormalization.__init__`, instead of subclassing it?

Reply:
I think it would be difficult to make `selector` a separate function, as the parameters used to select the backend are different for backend in `chainer`: (general, iDeep, cudnn) and backends in `chainermn`: (mpi, nccl). As shown here we need to access `self.comm` to create the backends, which only exists in `chainermn`. Currently, with as a member of the `BatchNormalization`, `selector` can access all the members via `self`, which hide such difference. If we make `selector` a function, then we need to call it with all the parameters needed, i.e. having `self.comm` even in BN in `chainer`. 
Another option is to pass the `BatchNormalization` object itself to the `selector`, but it kinda creates a loop like:
`self.bn_backend = self.bn_selector(self)`

After discussing with @shu65, we adopted this approach.

Reply:
Thanks.
`comm` is available at the moment of `MultiNodeBatchNormalizationFunction.__init__()` call.
I think you could write this way:

```py
def _create_selector(comm):
    def selector(other parameters...):
        if ...:
            return _NcclBackend(comm)
        ...
    return selector

func = BatchNormalization(..., _create_selector(comm))
```

`_create_selector` could also be replaced with a callable class.

It might look tricky, but I think the whole architecture is simpler because we won't need inheritance.

Reply:
I made a draft as 36b1549, can you see if you like it?